modules/videosegmenter-ffmpeg/src/main/java/org/opencastproject/videosegmenter/ffmpeg/VideoSegmenterServiceImpl.java

```diff
@@ -379,252 +453,506 @@ protected Catalog segment(Job job, Track track)
-            mpeg7, segments);
-        // filtered
-        LinkedList<Segment> segmentsNew = new LinkedList<Segment>();
-        OptimizationStep currentStepFiltered = new OptimizationStep(
-                changesThresholdLocal, 0,
-                prefNumberLocal, filterSegmentation(segments, track, segmentsNew, stabilityThreshold * 1000), segments);
-        currentStepFiltered.setSegmentNumAndRecalcErrors(segmentsNew.size());
-
-        logger.info("Segmentation yields {} segments after filtering", segmentsNew.size());
-
-        OptimizationStep currentStepBest;
-
-        // save better optimization in optimizationList
-        //
-        // the unfiltered segmentation is better if
-        // - the error is smaller than the error of the filtered segmentation
-        // OR - the filtered number of segments is smaller than the preferred number
-        //    - and the unfiltered number of segments is bigger than a value that should roughly estimate how many
-        //          segments with the length of the stability threshold could maximally be in a video
-        //          (this is to make sure that if there are e.g. 1000 segments and the filtering would yield
-        //           smaller and smaller results, the stability threshold won't be optimized in the wrong direction)
-        //    - and the filtered segmentation is not already better than the maximum error
-        if (currentStep.getErrorAbs() <= currentStepFiltered.getErrorAbs() || (segmentsNew.size() < prefNumberLocal
-                && currentStep.getSegmentNum() > (track.getDuration() / 1000.0f) / (stabilityThreshold / 2)
-                && !(currentStepFiltered.getErrorAbs() <= maxError))) {
-
-          optimizationList.add(currentStep);
-          Collections.sort(optimizationList);
-          currentStepBest = currentStep;
-          unusedResultsList.add(currentStepFiltered);
-        } else {
-          optimizationList.add(currentStepFiltered);
-          Collections.sort(optimizationList);
-          currentStepBest = currentStepFiltered;
+  /**
+   * Parses Chapter information out of a FFMETADATA file (see https://ffmpeg.org/ffmpeg-formats.html section 5 )
+   * @param reader The Reader to parse from
+   * @return List of parsed chapters
+   * @throws IOException When the reading the reader fails
+   * @throws ParseException When the parsing of the FFMETADATA reader fails
+   */
+  private List<Chapter> parseChapter(final BufferedReader reader) throws IOException, ParseException {
+    List<Chapter> chapters = new ArrayList<Chapter>();
+
+    int state = 0;
+    // Nanoseconds are the default timebase
+    final double defaultTimebase = 1e-9f;
+    double timebase = defaultTimebase;
+    long start = -1;
+    long end = -1;
+    Optional<StringBuilder> title = Optional.empty();
+
+    String line = reader.readLine();
+    int lineNumber = 1;
+    while (null != line) {
```

Comment:
A similar problem exists with state 5 as it can now directly lead into the new state 7. Additionally a ParseException would probably be better instead of a NullPointerException for unexpected file ends.

Reply:
Will push an update.

Reply:
Hopefully this fixes the problem: [55163d..bd6b50](https://github.com/opencast/opencast/compare/55163d681e85628a8645d636adfd00dd3cd44321..bd6b50c24c2372e8c20f6d2234dd03fb32fdee84)

Reply:
This opens up the possibility of `line` being `null` during the loop, which can now cause NullPointerExceptions with e.g. `startsWith("")`

Reply:
This only affects the comments/empty line if-branch or did I overlook something else?
[Fix](https://github.com/opencast/opencast/compare/bd6b50c24c2372e8c20f6d2234dd03fb32fdee84..6cd02b092cc1286f3b9d13ee9b2293f3f8f7a68b)

Reply:
Seems so, I just saw that it would be a problem in at least one case ^^.